### PR TITLE
refactor!: remove QuantumState classes

### PR DESCRIPTION
**Pull request by spflueger**
This PR 
- removes the `QuantumState` and `ComplexEnergyState` classes which will not be used in this form internally
- merges  the `ComplexEnergy` functionality with `Particle` and flattens the quantum number structure in `Particle`

This influences issue #231 and should be easier to tackle

### CODE REVIEW

File: expertsystem/data.py [hunk 1]
```diff
@@ -144,103 +137,31 @@
...
-
-    def __init__(  # pylint: disable=too-many-arguments
-        self,
-        name: str,
-        pid: int,
-        state: QuantumState[float],
-        mass: float,
-        width: float = 0.0,
-    ):
+    def __post_init__(self) -> None:
```
Comment (redeboer) [inline expertsystem/data.py:148]: Ok this is nice :+1:

File: expertsystem/data.py [hunk 2]
```diff
@@ -3,11 +3,8 @@
 __all__ = [  # fix order in API
```
Comment (redeboer) [inline expertsystem/data.py:3]: This was added in order to (1) hide certain methods from the API and (2) fix the order of the definitions. It seems, however, that (2) fails somehow:
file:///mnt/c/dev/PWA/ComPWA/expertsystem/doc/_build/html/api/expertsystem.data.html#module-expertsystem.data

Not sure why it doesn't work anymore since 11d36d37e028bdd6faf922cec741de76db7a1c3f (upgrade from Sphinx 2.4.4 to 3.2.1), see also "order" in
https://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html#directives

File: tests/channels/test_d0_to_ks_kp_km.py [hunk 3]
```diff
@@ -20,7 +20,7 @@ def test_script():
             "phi(1020)",
         ],
     )
-    stm.number_of_threads = 1
+    stm.number_of_threads = 2
```
Comment (redeboer) [inline tests/channels/test_d0_to_ks_kp_km.py:23]: Why is this changed to 2 again? Afaik, this affects test coverage, see f99cfeab17f3449ae94a7e1af8220826182ecf9f

File: tests/unit/io/test_pdg.py [hunk 4]
```diff
@@ -53,8 +53,4 @@ def test_pdg_entries(pdg, particle_database):
             continue
         internal_particle = particle_database[name]
         pdg_particle = pdg[name]
-        assert pdg_particle.name == internal_particle.name
-        assert pdg_particle.pid == internal_particle.pid
-        assert pdg_particle.state == internal_particle.state
-        assert pdg_particle.mass == pytest.approx(internal_particle.mass)
-        assert pdg_particle.width == pytest.approx(internal_particle.width)
+        assert pdg_particle == internal_particle
```
Comment (redeboer) [inline tests/unit/io/test_pdg.py:56]: I wonder if this test still makes sense. It was there at the time when `particle_list.*ml` was still there.

**Review by redeboer — COMMENTED** [review]
I quite like the consequences for usage (that is, removal of `state` from `Particle`), although this wasn't the motivation for this PR. For sure the removal of `ComplexEnergyState` makes the code easier to understand.

Just a few comments, but in principle I agree with the changes.

File: expertsystem/data.py [hunk 5]
```diff
@@ -3,11 +3,8 @@
...
     "ParticleCollection",
     "Particle",
-    "ComplexEnergyState",
-    "QuantumState",
     "Parity",
     "Spin",
-    "ComplexEnergy",
     "create_antiparticle",
     "create_particle",
     "GellmannNishijima",
```
Comment (redeboer) [inline expertsystem/data.py:10]: This was added in order to (1) hide certain methods from the API and (2) fix the order of the definitions. It seems, however, that (2) fails somehow:
https://pwa--244.org.readthedocs.build/projects/expertsystem/en/244/api/expertsystem.data.html#module-expertsystem.data

Not sure why it doesn't work anymore since 11d36d37e028bdd6faf922cec741de76db7a1c3f (upgrade from Sphinx 2.4.4 to 3.2.1), see also "order" in
https://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html#directives

File: tests/channels/test_d0_to_ks_kp_km.py (hunk 3 above)
Comment (spflueger) [reply tests/channels/test_d0_to_ks_kp_km.py:23]: Ah I forgot about the test coverage issue... I was just a remainder from pickling checks i did. I will change it back

File: tests/unit/io/test_pdg.py (hunk 4 above)
Comment (spflueger) [reply tests/unit/io/test_pdg.py:56]: indeed :smile:, i will remove it

File: expertsystem/data.py [hunk 6]
```diff
@@ -116,20 +111,18 @@
...
-    This is to make spin projection required in `.QuantumState` and unavailable
-    in `.Particle`.
+    Can **only** contain info that the `PDG <http://pdg.lbl.gov/>`_ would list.
+    Particles do NOT contain spin projection information.
     """
 
-    spin: _T
+    name: str
+    pid: int
+    energy: complex
```
[the remark this answers is not in the document]
Comment (spflueger) [reply expertsystem/data.py:124]: Hmm, you know me :smirk: . Well I will try to make the interface a bit more flexible (hope that is possible with the dataclasses)

File: expertsystem/data.py [hunk 7]
```diff
@@ -116,20 +111,18 @@
...
-    Generic[_T]
-):  # pylint: disable=too-many-instance-attributes
-    """Set of quantum numbers with a **generic type spin**.
+class Particle:  # pylint: disable=too-many-instance-attributes
+    """Immutable container of data defining a physical particle.
 
-    This is to make spin projection required in `.QuantumState` and unavailable
-    in `.Particle`.
+    Can **only** contain info that the `PDG <http://pdg.lbl.gov/>`_ would list.
+    Particles do NOT contain spin projection information.
```
[the remark this answers is not in the document]
Comment (spflueger) [reply expertsystem/data.py]: 5e108b938402e1f7e0f229f1ee9b0a157c419be4

File: expertsystem/data.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (spflueger) [reply expertsystem/data.py]: 5e108b938402e1f7e0f229f1ee9b0a157c419be4
Unfortunately its not possible to have energy and mass + width as two alternative init ways. I went back to mass + width then

File: tests/channels/test_d0_to_ks_kp_km.py (hunk 3 above)
Comment (spflueger) [reply tests/channels/test_d0_to_ks_kp_km.py]: 5e108b938402e1f7e0f229f1ee9b0a157c419be4

File: tests/unit/io/test_pdg.py (hunk 4 above)
Comment (spflueger) [reply tests/unit/io/test_pdg.py]: 5e108b938402e1f7e0f229f1ee9b0a157c419be4

File: expertsystem/data.py (hunk 5 above)
Comment (spflueger) [reply expertsystem/data.py:10]: I see, can you take a look at this?

Comment (redeboer) [reply expertsystem/data.py:10]: Yeah I checked after writing, don't have a good solution for it now. As this problem was there already before this PR, let's leave it as is.

**Review by redeboer — APPROVED** [review]
Good job :+1:
